packages/core/src/operator/operator-runtime/listenForChallengesCallback.ts

```diff
@@ -0,0 +1,114 @@
+ * @param {NodeLicenseStatus | string} newStatus - The new status to display
+ */
+export async function listenForChallengesCallback(challengeNumber: bigint, challenge: Challenge, event?: any) {
+
+    if (event && challenge.rollupUsed === config.rollupAddress) {
+        compareWithCDN(challenge)
+            .then(({ publicNodeBucket, error }) => {
+                if (error) {
+                    operatorState.onAssertionMissMatchCb(publicNodeBucket, challenge, error);
+                    return;
+                }
+                operatorState.cachedLogger(`Comparison between PublicNode and Challenger was successful.`);
+            })
+            .catch(error => {
+                operatorState.cachedLogger(`Error on CND check for challenge ${Number(challenge.assertionId)}.`);
+                operatorState.cachedLogger(`${error.message}.`);
+            });
+    }
+
+    operatorState.cachedLogger(`Received new challenge with number: ${challengeNumber}.`);
+    operatorState.cachedLogger(`Processing challenge...`);
+
+    // Add a delay of 1 -300 seconds to the new challenge process so not all operators request the subgraph at the same time
+    const delay = Math.floor(Math.random() * 301);
+    await new Promise((resolve) => {
+        setTimeout(resolve, delay * 1000);
+    })
+
+    try {
+        const graphStatus = await getSubgraphHealthStatus();
+        if (graphStatus.healthy) {
+            const { sentryWalletMap, sentryKeysMap, nodeLicenseIds, mappedPools, refereeConfig } =
+                await loadOperatorKeysFromGraph(operatorState.operatorAddress, challengeNumber - 1n);
+
+            await processNewChallenge(challengeNumber, challenge, nodeLicenseIds, sentryKeysMap, sentryWalletMap, mappedPools, refereeConfig);
+            // check the previous challenge, that should be closed now
+            if (challengeNumber > BigInt(1)) {
+                await processClosedChallenges(challengeNumber - BigInt(1), nodeLicenseIds, sentryKeysMap, sentryWalletMap);
+            }
+
+
+        } else {
+            operatorState.cachedLogger(`Revert to RPC call instead of using subgraph. Subgraph status error: ${graphStatus.error}`)
+
+            const { sentryKeysMap, nodeLicenseIds } = await loadOperatorKeysFromRPC(operatorState.operatorAddress);
+
+            await processNewChallenge(challengeNumber, challenge, nodeLicenseIds, sentryKeysMap);
+            // check the previous challenge, that should be closed now
+            if (challengeNumber > BigInt(1)) {
+                await processClosedChallenges(challengeNumber - BigInt(1), nodeLicenseIds, sentryKeysMap);
+            }
+        }
+    } catch (error: any) {
+        operatorState.cachedLogger(`Error processing new challenge in listener callback: - ${error && error.message ? error.message : error}`);
+    }
+}
+
+// Helper function to request the assertion from the public node's CDN
+async function getPublicNodeFromBucket(confirmHash: string) {
```

Comment:
Should this be in a separate file in utils?

Reply:
Same as above :)